File: C++/Queue/Queue_04.cpp
/*
===============================================================================
Program Name    : Queue Count Operation
Program Number  : 004
File Name       : Queue_04.cpp

Description     : This program implements the Count operation of a Queue
                  using a singly linear linked list. The Count operation
                  returns the total number of elements currently present
                  in the Queue.

Date            : August 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

class Queue
{
private:
    struct node *first;
    int iCount;

public:
    Queue();
    void Enqueue(int iNo);
    int Dequeue();
    void Display();
    int Count();
};

Queue :: Queue()
{
    this->first = NULL;
    this->iCount = 0;
}

void Queue :: Enqueue(int iNo)
{
    struct node *newn = NULL;
    struct node *temp = NULL;

    newn = new struct node();

    newn->data = iNo;
    newn->next = NULL;

    if(first == NULL)
    {
        first = newn;
    }
    else
    {
        temp = first;

        while(temp->next != NULL)
        {
            temp = temp->next;
        }

        temp->next = newn;
    }

    iCount++;
}

void Queue :: Display()
{
    struct node *temp = NULL;

    temp = first;

    while(temp != NULL)
    {
        cout<<"| "<<temp->data<<" |"<<endl;
        temp = temp->next;
    }
}

int Queue :: Count()
{
    return iCount;
}

int Queue :: Dequeue()
{
    return 0;
}

int main()
{
    Queue qobj;
    int iRet = 0;

    qobj.Enqueue(11);
    qobj.Enqueue(21);
    qobj.Enqueue(51);
    qobj.Enqueue(101);

    qobj.Display();

    iRet = qobj.Count();

    cout<<"\nNumber of Elements in Queue are : "<<iRet<<endl;

    return 0;
}
/*
===============================================================================
Expected Output
===============================================================================

| 11 |
| 21 |
| 51 |
| 101 |

Number of Elements in Queue are : 4

===============================================================================
Algorithm : Count Queue Elements
===============================================================================

1. Start the program.

2. Create a Queue object.

3. Insert elements into the Queue using Enqueue().

4. Display the Queue elements.

5. Call the Count() function.

6. Return the value of iCount.

7. Display the number of elements present in the Queue.

8. Stop the program.

===============================================================================
Time Complexity
===============================================================================

Enqueue Operation : O(n)

Display Operation : O(n)

Count Operation   : O(1)

Best Case         : O(1)
Average Case      : O(1)
Worst Case        : O(1)

Space Complexity  : O(1) auxiliary space

===============================================================================
Conclusion
===============================================================================

This program implements the Count operation of a Queue using a singly
linear linked list. The iCount data member keeps track of the number of
elements inserted into the Queue. The Count() function directly returns
this value, making the operation efficient with O(1) time complexity.

===============================================================================
*/

File: Java/Queue/Queue_04.java
/*
===============================================================================
Program Name    : Queue - Count Operation
Program Number  : 04
File Name       : Queue_04.java

Description     : This program implements the Count operation of a Queue
                  using a singly linear linked list. The Count operation
                  returns the total number of elements currently present
                  in the Queue.

Date            : August 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class
2. Object-Oriented Programming (OOP)
3. Self-Referential Class
4. Constructor
5. Singly Linear Linked List
6. Queue
7. Enqueue Operation
8. Display Operation
9. Count Operation
10. Instance Variables

Key Learning Points:
- Understand the basic structure of a Queue in Java.
- Learn how a Queue is represented using a linked list.
- Understand insertion of elements at the last position.
- Learn how iCount keeps track of Queue elements.
- Understand how the Count operation returns the number of elements.
- Learn how to display Queue elements.

===============================================================================
*/

class node
{
    public int data;
    public node next;
}

class Queue
{
    private node first;
    private int iCount;

    public Queue()
    {
        first = null;
        iCount = 0;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Enqueue
    // Description   : Insert element at Last
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Enqueue(int iNo)
    {
        node newn = null;
        node temp = null;

        newn = new node();

        newn.data = iNo;
        newn.next = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            temp = first;

            while(temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
        }

        iCount++;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Display
    // Description   : Display Queue elements
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Display()
    {
        node temp = first;

        System.out.println("\nElements of Queue are : ");

        while(temp != null)
        {
            System.out.print("|"+temp.data+"| -> ");
            temp = temp.next;
        }

        System.out.println("NULL");
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Count
    // Description   : Count number of elements
    //
    ///////////////////////////////////////////////////////////////////////////

    public int Count()
    {
        return iCount;
    }
}

class Queue_04
{
    public static void main(String A[])
    {
        Queue qobj = new Queue();

        int iRet = 0;

        qobj.Enqueue(11);
        qobj.Enqueue(21);
        qobj.Enqueue(51);
        qobj.Enqueue(101);

        qobj.Display();

        iRet = qobj.Count();

        System.out.println("\nNumber of Elements : "+iRet);
    }
}

/*
===============================================================================
Expected Output:
===============================================================================

Elements of Queue are :
|11| -> |21| -> |51| -> |101| -> NULL

Number of Elements : 4

===============================================================================
Memory Representation:
===============================================================================

                         Queue Object

                 +-----------------------+
                 | first -----------------|------+
                 | iCount = 4             |      |
                 +-----------------------+      |
                                                |
                                                v
                                      +------------------+
                                      | data = 11        |
                                      | next ------------|------+
                                      +------------------+      |
                                                                |
                                                                v
                                                      +------------------+
                                                      | data = 21        |
                                                      | next ------------|------+
                                                      +------------------+      |
                                                                                |
                                                                                v
                                                                      +------------------+
                                                                      | data = 51        |
                                                                      | next ------------|------+
                                                                      +------------------+      |
                                                                                              |
                                                                                              v
                                                                                    +------------------+
                                                                                    | data = 101       |
                                                                                    | next = NULL      |
                                                                                    +------------------+

===============================================================================
Algorithm:
===============================================================================

1. Start the program.
2. Create a Queue object.
3. Initialize first to NULL and iCount to 0.
4. Insert 11, 21, 51 and 101 using Enqueue().
5. Create a new node for each element.
6. Insert every new node at the last position of the Queue.
7. Increment iCount after every successful insertion.
8. Call Display() to display all Queue elements.
9. Call Count() to get the total number of elements.
10. Store the returned value in iRet.
11. Display the number of elements.
12. Stop the program.

===============================================================================
Program Flow:
===============================================================================

        Start
          |
          v
    Create Queue Object
          |
          v
   Initialize Queue
          |
          v
    Enqueue Elements
          |
          v
     Display Queue
          |
          v
       Call Count()
          |
          v
   Display Element Count
          |
          v
         Stop

===============================================================================
Time Complexity:
===============================================================================

Enqueue() : O(n)
Display() : O(n)
Count()   : O(1)

===============================================================================
Space Complexity:
===============================================================================

O(n) total space for n Queue nodes.

===============================================================================
Conclusion:
===============================================================================

Program 04 demonstrates the Count operation of a Queue using a singly
linear linked list in Java. The iCount variable is incremented whenever
a new element is inserted into the Queue. The Count() function returns
the current number of elements in constant time.

===============================================================================
*/
